hai/pipe_pump.py

```diff
@@ -27,20 +27,31 @@ def register(self, key: str, fileobj: Optional[IO[bytes]]) -> None:
         if fileobj:
             self.selector.register(fileobj, selectors.EVENT_READ, data=key)
 
-    def pump(self, timeout: float = 0) -> None:
+    def pump(self, timeout: float = 0, repeat: int = 1) -> int:
         """
         Pump the pipes and buffer incoming data.
 
+        Returns the number of times reads occurred.
+
         :param timeout: Timeout for reading.
                         0 = just poll and return immediately.
+        :param repeat: Maximum times to repeat reading until nothing is read anymore.
+        :return: The number of reads done.
         """
         with self.selector_lock:
             if not self.selector:  # pragma: no cover
-                return
-            for (key, _event) in self.selector.select(timeout=timeout):
-                fileobj = key.fileobj  # type: IO[bytes]  # type: ignore[assignment]
-                data = fileobj.read(self.read_size)
-                self.feed(key.data, data)
+                return 0
+            for reads in range(repeat):
+                should_repeat = False
+                for (key, _event) in self.selector.select(timeout=timeout):
+                    fileobj = key.fileobj  # type: IO[bytes]  # type: ignore[assignment]
+                    data = fileobj.read(self.read_size)
+                    self.feed(key.data, data)
+                    if repeat:
```

Comment:
shouldn't this be `if reads` as `repeat` is the function parameter which should always be truelike

Reply:
or, no, `range(1)` would give it 0 hmm

Reply:
a bit confusing logic :thinking:

Reply:
I agree. I might revisit it...